src/java/org/apache/cassandra/cache/ChunkCache.java

```diff
@@ -310,29 +316,53 @@ public CachingRebufferer(ChunkReader file)
         }
 
         @Override
-        public Buffer rebuffer(long position)
+        public BufferHolder rebuffer(long position)
         {
-            int spin = 0;
             try
             {
                 long pageAlignedPos = position & alignmentMask;
-                Buffer buf;
-                Key key = new Key(source, internedPath, pageAlignedPos);
+                BufferHolder buf;
+                Key chunkKey = new Key(source, internedPath, pageAlignedPos);
+
+                int spin = 0;
+                //There is a small window when a released buffer/invalidated chunk
+                //is still in the cache. In this case it will return null
+                //so we spin loop while waiting for the cache to re-populate
                 while (true)
                 {
-                    buf = cache.get(key).reference();
+                    Buffer chunk;
+                    // Using cache.get(k, compute) results in lots of allocation, rather risk the unlikely race...
+                    CompletableFuture<Buffer> cachedValue = cache.getIfPresent(chunkKey);
+                    if (cachedValue == null)
+                    {
+                        // this blocking read might compete with an another read, but the race is benign (under the assumption that caching is valid)
+                        CompletableFuture<Buffer> entry = new CompletableFuture<>();
+                        // Put the future in the cache first so the window of racing (and loading the same chunk twice) is small
+                        cache.put(chunkKey, entry);
+
+                        try
+                        {
+                            chunk = load(chunkKey);
+                        }
+                        catch (Throwable t)
+                        {
+                            // also signal other waiting readers
```

Comment:
might be worth noting this also cleans up the cache

Reply:
How does it clean the cache? I'm missing that.

It marks the future as invalid and the other readers would get an exception if they ever read the same key. 
But the entry (the future) won't be immediately removed from the cache. 
I wonder if this is the behavior we really want. Failing to load a chunk will cache the failure and won't reattempt to load the chunk until the chunk gets evicted by the cache eviction policy.

Reply:
It’s not obvious but Caffeine adds a whenComplete action to remove a failed future or update the metadata if successful. A negative cache should be a success with the result indicating that, e.g. caching an Optional vs null in a synchronous cache. I think that’s what @michaeljmarshall realized so a note communicates that implementation detail.

Reply:
I have added a test to cover this case and demonstrate this behavior